aws-lc-rs/src/kem.rs

```diff
@@ -0,0 +1,468 @@
+            let mut secret_key_size = alg.secret_key_size();
+            let mut priv_key_bytes = vec![0u8; secret_key_size];
+            if 1 != EVP_PKEY_get_raw_private_key(
+                kyber_key.as_const_ptr(),
+                priv_key_bytes.as_mut_ptr(),
+                &mut secret_key_size,
+            ) {
+                return Err(Unspecified);
+            }
+            Ok(KemPrivateKey {
+                algorithm: alg,
+                pkey: kyber_key,
+                priv_key: priv_key_bytes.into(),
+            })
+        }
+    }
+
+    /// Return the algorithm associated with the given KEM private key.
+    #[must_use]
+    pub fn algorithm(&self) -> &'static KemAlgorithm {
+        self.algorithm
+    }
+
+    /// Computes the KEM public key from the KEM private key
+    ///
+    /// # Errors
+    /// `error::Unspecified` when operation fails due to internal error.
+    ///
+    pub fn compute_public_key(&self) -> Result<KemPublicKey, Unspecified> {
+        let mut public_key_size = self.algorithm.public_key_size();
+        let mut pubkey_bytes = vec![0u8; public_key_size];
+        unsafe {
+            if 1 != EVP_PKEY_get_raw_public_key(
+                self.pkey.as_const_ptr(),
+                pubkey_bytes.as_mut_ptr(),
+                &mut public_key_size,
+            ) {
+                return Err(Unspecified);
+            }
+
+            let pubkey = LcPtr::new(EVP_PKEY_kem_new_raw_public_key(
+                self.algorithm.id.nid(),
+                pubkey_bytes.as_ptr(),
+                public_key_size,
+            ))?;
+
+            Ok(KemPublicKey {
+                algorithm: self.algorithm,
+                pkey: pubkey,
+                pub_key: pubkey_bytes.into(),
+            })
+        }
+    }
+
+    /// Performs the decapsulate operation using the current KEM private key on the given ciphertext.
+    ///
+    /// `ciphertext` is the ciphertext generated by the encapsulate operation using the KEM public key
+    /// associated with the current KEM private key.
+    ///
+    /// After the decapsulation is finished, `decapsulate` calls `kdf` with the raw shared secret
+    /// from the operation and then returns what `kdf` returns.
+    ///
+    /// # Errors
+    /// `error_value` when operation fails due to internal error.
+    ///
+    #[allow(clippy::missing_panics_doc)]
+    pub fn decapsulate<F, R, E>(
+        &self,
+        ciphertext: &mut [u8],
+        error_value: E,
+        kdf: F,
+    ) -> Result<R, E>
+    where
+        F: FnOnce(&[u8]) -> Result<R, E>,
+    {
+        unsafe {
+            let ctx = LcPtr::new(EVP_PKEY_CTX_new(*self.pkey, null_mut()));
+            if ctx.is_err() {
+                return Err(error_value);
+            }
+            let mut shared_secret_len = self.algorithm.shared_secret_size();
+            let mut shared_secret: Vec<u8> = vec![0u8; shared_secret_len];
+
+            if 1 != EVP_PKEY_decapsulate(
+                *ctx.unwrap(),
```

Comment:
Avoid the unwrap panic here. Instead prefer something like `.map_err(|_|Unspecified)?`

Reply:
This will never fail because we check if ctx is an error. `map_err` just maps the error, it doesn't actually get the value. We can't use `?` because the error type is caller defined.

Reply:
Ben - I see what you mean, but stylistically, we want to avoid using `unwrap` (as it's a "code smell"). The problem with `map_err` is that we need to return `error_value` but the type `E` is not `Copy` so it would get "moved" into the closure, and the borrow checker wouldn't let it be used later in the function.

The best way might be to use a let-else for the assignment so that `ctx` gets the `LcPtr`, the borrow-checker is happy, and we don't have to call `unwrap`:
```
    let Ok(ctx) =
        LcPtr::new(EVP_PKEY_CTX_new(*self.pkey, null_mut())) else {
        return Err(error_value);
    };
```

Reply:
Yeah the let else looks clean. [agreement.rs](https://github.com/aws/aws-lc-rs/blob/main/aws-lc-rs/src/agreement.rs#L534) uses unwrap in the exact same way so we should probably refactor there then too.

Reply:
> The problem with map_err is that we need to return error_value but the type E is not Copy so it would get "moved" into the closure, and the borrow checker wouldn't let it be used later in the function.

Right....I see what you mean about the "move".

Reply:
> Yeah the `let else` looks clean. [Agreement.rs](https://github.com/aws/aws-lc-rs/blob/main/aws-lc-rs/src/agreement.rs#L534) uses unwrap in the exact same way, so we should probably refactor there then too.

https://github.com/aws/aws-lc-rs/pull/177